src/V3VirtIface.cpp

```diff
@@ -0,0 +1,175 @@
+//
+// Copyright 2003-2023 by Wilson Snyder. This program is free software; you
+// can redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+//*************************************************************************
+// V3VirtIface's Transformations:
+//
+// Each interface type written to via virtual interface, or written to normally but read via
+// virtual interface:
+//     Create a trigger var for it
+// Each AssignW:
+//     If it writes to a virtual interface, or to a variable read via virtual interface:
+//         Convert to an always
+// Each statement:
+//     If it writes to a virtual interface, or to a variable read via virtual interface:
+//         Set the corresponding trigger to 1
+//         If the write is done by an AssignDly, the trigger is also set by AssignDly
+//
+//*************************************************************************
+
+#include "V3PchAstNoMT.h"  // VL_MT_DISABLED_CODE_UNIT
+
+#include "V3VirtIface.h"
+
+#include "V3AstNodeExpr.h"
+#include "V3UniqueNames.h"
+
+VL_DEFINE_DEBUG_FUNCTIONS;
+
+//######################################################################
+//
+
+class VirtIfaceVisitor final : public VNVisitor {
+private:
+    // NODE STATE
+    // AstIface::user1() -> AstVarScope*. Trigger var for this interface
+    const VNUser1InUse m_user1InUse;
+
+    // TYPES
+    using OnWriteToVirtIface = std::function<void(AstVarRef*, AstIface*)>;
+
+    // STATE
+    AstNetlist* const m_netlistp;  // Root node
+    AstNodeAssign* m_trigAssignp = nullptr;  // Previous/current trigger assignment
+    AstIface* m_trigAssignIfacep = nullptr;  // Interface type whose trigger is assigned
+                                             // by m_trigAssignp
+    V3UniqueNames m_vifTriggerNames{"__VvifTrigger"};  // Unique names for virt iface
+                                                       // triggers
+
+    // METHODS
+    static void foreachWrittenVirtIface(AstNode* const nodep, const OnWriteToVirtIface& onWrite) {
+        nodep->foreach([&](AstVarRef* const refp) {
+            if (refp->access().isReadOnly()) return;
+            if (AstIfaceRefDType* dtypep = VN_CAST(refp->varp()->dtypep(), IfaceRefDType)) {
+                if (dtypep->isVirtual() && VN_IS(refp->firstAbovep(), MemberSel)) {
+                    onWrite(refp, dtypep->ifacep());
+                }
+            } else if (AstIface* ifacep = refp->varp()->sensIfacep()) {
+                onWrite(refp, ifacep);
+            }
+        });
+    }
+    static void unsupportedWriteToVirtIface(AstNode* nodep) {
+        if (!nodep) return;
+        foreachWrittenVirtIface(nodep, [](AstVarRef* const selp, AstIface*) {
+            selp->v3warn(E_UNSUPPORTED,
+                         "Unsupported: write to virtual interface in this location");
```

Comment:
Any other hints we might be able to provide e.g. under fork or something?  Or can we say e.g. write to virtual interface not under {something}?